### PR TITLE
Update conformer

### PR DESCRIPTION
**Pull request by VahidooX**
+ Polished the conformer code and configs.
+ Added mask_value to the specaug
+ Refactored positional embedding classes

### CODE REVIEW

File: nemo/collections/asr/parts/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/multi_head_attention.py:248]: Add else raise error case here.

File: nemo/collections/asr/parts/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/multi_head_attention.py:237]: Add else raise error case here.

File: nemo/collections/asr/parts/multi_head_attention.py
Comment (titu1994) [inline nemo/collections/asr/parts/multi_head_attention.py:223]: Constructor doesn't take pos_type argument ?

File: nemo/collections/asr/parts/conformer_modules.py
Comment (titu1994) [inline nemo/collections/asr/parts/conformer_modules.py:98]: Document differences between rel pos types

File: nemo/collections/asr/models/ctc_models.py
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:376]: Remove timing code.

File: nemo/collections/asr/models/ctc_models.py
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:400]: Remove debugging

File: nemo/collections/asr/models/ctc_models.py
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:405]: Remove logging

File: nemo/collections/asr/models/ctc_models.py
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:412]: This needs to be encoded_len same as else case

File: nemo/collections/asr/metrics/wer_bpe.py
Comment (titu1994) [inline nemo/collections/asr/metrics/wer_bpe.py:105]: Add type of prediction Len - torch.Tensor

File: nemo/collections/asr/metrics/wer.py
Comment (titu1994) [inline nemo/collections/asr/metrics/wer.py:142]: Add type to prediction_len - torch.Tensor

File: examples/asr/speech_to_text_rnnt_bpe.py
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt_bpe.py:82]: Use with open_dict() and pop out this argument - it should not remain inside the config if submitted by user.

File: examples/asr/speech_to_text_rnnt_bpe.py
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt_bpe.py:86]: Load checkpoint on seperate line, load weights and then explicitly del the entire checkpoint. Otherwise the optimizer state will stay over and won't get deleted before model creation + training begins.

File: examples/asr/speech_to_text_rnnt_bpe.py
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt_bpe.py:17]: Add from omegaconf import open_dict

File: examples/asr/speech_to_text_rnnt.py
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt.py:95]: Same comments as below - pop out this command line override, load checkpoint in seperate line, load checkpoint into model and explicitly del the entire checkpoint. 

X = torch.load()
Model.load_state_dict(X[state_dict])
del X

File: examples/asr/speech_to_text_rnnt.py
Comment (titu1994) [inline examples/asr/speech_to_text_rnnt.py:16]: Add from omegaconf import open_dict

File: examples/asr/speech_to_text_bpe.py
Comment (titu1994) [inline examples/asr/speech_to_text_bpe.py:76]: Same comments as below

Comment (titu1994) [reply examples/asr/speech_to_text_bpe.py:76]: + add this even to the speech_to_text.py script please. Quite convenient

**Review by titu1994 — CHANGES_REQUESTED** [review]
Overall looks good, minor comments and some refactoring needed. @okuchaiev could you review update to WER calculation - it will be masked by sequence length now (if the model provides it)

Finally, let's drop all the vocab. It adds space to Nemo for no reason and we don't want dozens of vocab for every single dataset. The model trained on any dataset contains the tokekizer and the vocab inside it so it is unnecessary to add it as a data file. (we need to find some external place to store it - maybe as a GitHub release? @okuchaiev)

File: nemo/collections/asr/metrics/wer.py
Comment (titu1994) [inline nemo/collections/asr/metrics/wer.py:117]: Add type to prediction_len

File: nemo/collections/asr/metrics/wer_bpe.py
Comment (titu1994) [inline nemo/collections/asr/metrics/wer_bpe.py:80]: Add type to prediction_len

File: nemo/collections/asr/models/ctc_models.py
Comment (titu1994) [inline nemo/collections/asr/models/ctc_models.py:189]: @okuchaiev @vsl9 Please check if this is fine.

**Review by titu1994 — COMMENTED** [review]
Overall looks great, I'll let @okuchaiev  and @vsl9 review the change to WER before we merge

**Comment by okuchaiev** [comment]
@vsl9 , @jbalam-nv  and @blisc could you please double-check changes to WER computation: ``nemo/collections/asr/metrics/wer.py``

File: nemo/collections/asr/models/ctc_models.py
Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:189]: Do all models return logits_len?

Comment (VahidooX) [reply nemo/collections/asr/models/ctc_models.py:189]: Yes, all CTC_Models return it here : 
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/models/ctc_models.py#L390

It can be None and when it is None, it is ignored during WER calculation. But our current CTC models actually return the correct lengths.

For example, Jasper/QN return the lengths which are given by encoder here: 
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/modules/conv_asr.py#L192

Previously we did not use the lenghts return by the encoder at all in our codes. This would be the first time to actually use those.

Comment (blisc) [reply nemo/collections/asr/models/ctc_models.py:189]: It seems fine then. Theoretically, the Jasper QuartzNet could make predictions past the current returned value due to the receptive field, but I don't think this makes a large practical difference

**Review by blisc — APPROVED** [review]
WER changes looks good to me

**Review by titu1994 — APPROVED** [review]
LGTM

**Pull request merged**
